### DISCUSSION TITLE
Rebuild same version

### DISCUSSION THREAD

**Discussion by n0ise9914** [openupm/openupm#134] [Q&A] [answered]
Is there any way to rebuild existing version (tag) ?
It's not an issue, but i didn't know where to ask!

**Comment by favoyang** [comment, marked as answer]
@n0ise9914,

As a practice of public registry (like npm.org), published content is treated as immutable. Republish same version is not allowed, because of developers may already relay on the existing build, there's no way to keep them informed.

For OpenUPM, the build pipeline will build all failed builds, but it will do nothing if detecting a successful build being re-tagged. So in general I suggest package owner bump version to fix bugs..

There're a few cases as exception. i.e. during setting up a new package something may go wrong completely. Or when maintaining a forked project that not willing to bump version. In such case you can delete wrong tag, then request on GitHub issues, and I will do it manually. And it may take up to 24h to clean the global cache of CDN.

**Comment by n0ise9914** [comment]
Yes as you noted i've forgot to add meta file for package.json, and this repos are from official upstream,s and changing version isn't a good idea.
So, Thank you, i'll send a request (after testing everything)

Some other famous package repositories like Jcenter have different strategy,
After publishing files/versions to bintray you have 180 days of grace to delete them.

Granting limited access to owner of packages (api to ci/cd pipeline) will make the process easier.
How about adding pre-release stage to pipeline/platform?

**Comment by n0ise9914** [comment]
I've removed all releases/tags from these two packages:
com.google.google-signin-unity
com.gameanalytics

Would you please invalidate the cache?
Thanks.

**Comment by favoyang** [comment]
> I've removed all releases/tags from these two packages:
> com.google.google-signin-unity
> com.gameanalytics
> 
> Would you please invalidate the cache?

The two packages has been swiped. You can add new tags now. The CI checks repository every 5 minutes, and it takes 1-3 mins to make a build. So you may wait for 5-8 minutes to see the result.

**Comment by favoyang** [comment]
> Some other famous package repositories like Jcenter have different strategy,
> After publishing files/versions to bintray you have 180 days of grace to delete them.

Npm.org also allow delete/deprecated packages in a time range, but you can not republish deleted version. I will check Jcenter policy later.

> Granting limited access to owner of packages (api to ci/cd pipeline) will make the process easier.
> How about adding pre-release stage to pipeline/platform?

I haven't get a clear idea about how to achieve this. A few points,
- The republish ability should be limited.
- The openupm system is non-authenticated. But this feature requires a way to authenticate user, perhaps limited to package owner and package hunter.

**Comment by n0ise9914** [comment]
![CpWz_205](https://user-images.githubusercontent.com/8354673/78136960-b6f9bf00-7439-11ea-8daa-dfde11f254c5.png)
According to date 2020-04-01, it's clear that it's replaced.

I've also seen the pipeline succeed at:
[pipeline buildId-3393](https://dev.azure.com/openupm/openupm/_build/results?buildId=3393&view=logs&s=96ac2280-8cb4-5df5-99de-dd2da759617d)
And here:
[com.google.google-signin-unity-1.0.4.tgz](https://openupm.sfo2.cdn.digitaloceanspaces.com/verdaccio/com.google.google-signin-unity/com.google.google-signin-unity-1.0.4.tgz)

But is this the intended behavior (duplicate results for same package)? (thought of sharing it with you)

Because the official's provide .unitypackage files, and are manually imported, some temporary files are getting automatically removed by unity.
But when we use package manager the package directory of project is immutable (so unity couldn't remove this files).
I've noticed that the official branches had seven trash meta files (cause we have warnings) and removed these unnecessary .meta files from the branch.

**I need you to invalidate the cache of this two packages one more time (to take effect)**, i promise this is the last time😂

**Comment by favoyang** [comment]
> But is this the intended behavior (duplicate results for same package)? (thought of sharing it with you)

The duplication is a caused by a bug of cached search result. See https://github.com/openupm/openupm-cli/issues/4. Delete `~/.openupm/com.openupm.all.json` file will fix it.

```
$ openupm search google
fast search endpoint is not available, using old search.
┌──────────────────────────────────────────────┬──────────────────────┬──────────────┬────────────┐
│ Name                                         │ Version              │ Author       │ Date       │
├──────────────────────────────────────────────┼──────────────────────┼──────────────┼────────────┤
│ com.google.google-signin-unity               │ 1.0.4                │ Arzhang Mos… │ 2020-04-01 │
│ Google Sign-In API plugin                    │                      │              │            │
└──────────────────────────────────────────────┴──────────────────────┴──────────────┴────────────┘
```

I'm not sure what the issue of meta file, but com.google.google-signin-unity has been swiped. Of course you're welcome to test the upm branch before tagging it.
